src/helpers/web.ts

```diff
@@ -162,41 +168,62 @@ export function generateRequestHeadersPOST(
   token: string,
   query: string,
   source: string,
+  envs: UploaderEnvs,
   args: UploaderArgs,
 ): IRequestHeaders {
   const url = new URL(`upload/v4?package=${getPackage(
     source,
   )}&token=${token}&${query}`, postURL)
 
+  const initialHeaders = {
+    'X-Upload-Token': token,
+    'X-Reduced-Redundancy': 'false',
+  }
+
+  const headers: IncomingHttpHeaders = generateHeaders(envs, initialHeaders)
+
   return {
-    agent: args.upstream ? new ProxyAgent(args.upstream) : undefined,
+    agent: addProxyIfNeeded(envs, args),
     url: url,
     options: {
-      headers: {
-        'X-Upload-Token': token,
-        'X-Reduced-Redundancy': 'false',
-      },
+      headers,
       method: 'POST',
       origin: postURL,
       path: `${url.pathname}${url.search}`,
     },
   }
 }
 
+export function generateHeaders(envs: UploaderEnvs, initialHeaders: IncomingHttpHeaders): IncomingHttpHeaders {
```

Comment:
I'm not convinced this needs to be it's own function. It feels particularly dirty since the function is called `generateHeaders`, but we're giving it headers already. I think we should deal with all header generation here.

Reply:
I didn't want a code branch here, since POST and PUT use different headers.

Reply:
could we just name it something like `addProxyHeaders` then?